### PR TITLE
Additional convergence criterion

### PR DESCRIPTION
**Pull request by odespard**
At present, optimizers need to have reached the target number of precursors before they can converge. This poses a problem if the optimizer tries a value so narrow that it is difficult to identify the target number. However, such values are not likely to optimize the feature value (indeed, if the feature value is well chosen they shouldn't, as we should not pick parameter values that reduce precursor counts so much). This PR introduces a new set of methods which record when optimization has been skipped and considers optimization to have converged if the optimizer is skipped too many times (as defined by max_skips in the calibration field of the config). The max_skips field has a default value of 1 and is unlikely to be worth changing as long as we use an exponential batch plan, since skipping an optimizer more than once implies that performance in precursors identifications is at least roughly twice as bad as before.

### CODE REVIEW

File: alphadia/workflow/optimization.py [hunk 1]
```diff
@@ -49,16 +49,21 @@ def step(self, precursors_df: pd.DataFrame, fragments_df: pd.DataFrame):
...
         fragments_df: pd.DataFrame
-            The fragment dataframe for the search
+            The filtered fragment dataframe for the search (see peptidecentric.PeptideCentricWorkflow.filter_dfs).
 
 
         """
 
         pass
 
+    @abstractmethod
```
Comment (mschwoer) [inline alphadia/workflow/optimization.py:62]: you could just make this an ordinary method, and change the docstring to `Default implementation: skipping of optimization is not recorded`
(and remove it from the subclass where it's implemented as `pass`)

File: alphadia/workflow/optimization.py [hunk 2]
```diff
@@ -88,6 +93,7 @@ def __init__(
         self.workflow.optimization_manager.fit({self.parameter_name: initial_parameter})
         self.has_converged = False
         self.num_prev_optimizations = 0
+        self.num_consecutive_skips = 0
```
Comment (mschwoer) [inline alphadia/workflow/optimization.py:96]: could this (and others?) be private?

File: alphadia/workflow/optimization.py [hunk 3]
```diff
@@ -165,6 +157,17 @@ def step(
                 verbosity="progress",
             )
 
+    def skip(self):
+        """Increments the internal counter for the number of consecutive skips and checks if the optimization should be stopped."""
+        self.num_consecutive_skips += 1
+        if self._unnecessary_to_continue:
```
Comment (mschwoer) [inline alphadia/workflow/optimization.py:163]: `_unnecessary_to_continue` is a bit sloppy, mb find a better name for this concept?

**Review by mschwoer — APPROVED** [review]

File: alphadia/workflow/optimization.py (hunk 2 above)
Comment (odespard) [reply alphadia/workflow/optimization.py:96]: I do refer to them outside the class, but only for logging purposes. Should I just move the logging inside the class?

File: alphadia/workflow/optimization.py (hunk 1 above)
Comment (odespard) [reply alphadia/workflow/optimization.py:62]: Good idea.

File: alphadia/workflow/optimization.py (hunk 3 above)
Comment (odespard) [reply alphadia/workflow/optimization.py:163]: Yes, I agree. I tried for a while and couldn't come up with anything, but I will try again.

Maybe ``_batch_substantially_bigger``?

Comment (mschwoer) [reply alphadia/workflow/optimization.py:163]: mb also add some of the PR description here in order to fully get this

Comment (odespard) [reply alphadia/workflow/optimization.py:163]: Or ``_large_drop_in_precursor_proportion``?

File: alphadia/workflow/optimization.py (hunk 2 above)
Comment (odespard) [reply alphadia/workflow/optimization.py:96]: I have done this.

**Pull request merged**